external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala

```diff
@@ -0,0 +1,474 @@
+    val newPartitionOffsets = if (newPartitions.nonEmpty) {
+      fetchNewPartitionEarliestOffsets(newPartitions.toSeq)
+    } else {
+      Map.empty[TopicPartition, Long]
+    }
+
+    // Sort the partitions and current list of executors to consistently assign each partition
+    // to the executor. This allows cached KafkaConsumers in the executors to be re-used to
+    // read the same partition in every batch.
+    val topicPartitionOrdering = new Ordering[TopicPartition] {
+      override def compare(l: TopicPartition, r: TopicPartition): Int = {
+        implicitly[Ordering[(String, Long)]].compare(
+          (l.topic, l.partition),
+          (r.topic, r.partition))
+      }
+    }
+    val sortedTopicPartitions = untilPartitionOffsets.keySet.toSeq.sorted(topicPartitionOrdering)
+    val sortedExecutors = getSortedExecutorList(sc)
+    val numExecutors = sortedExecutors.size
+    logDebug("Sorted executors: " + sortedExecutors.mkString(", "))
+    val offsetRanges = sortedTopicPartitions.map { tp =>
+      val fromOffset = fromPartitionOffsets.get(tp).getOrElse {
+        newPartitionOffsets.getOrElse(tp, {
+          // This should not happen since newPartitionOffsets contains all paritions not in
+          // fromPartitionOffsets
+          throw new IllegalStateException(s"$tp doesn't have a offset")
+        })
+      }
+      val untilOffset = untilPartitionOffsets(tp)
+      val preferredLoc = if (numExecutors > 0) {
+        Some(sortedExecutors(positiveMod(tp.hashCode, numExecutors)))
+      } else None
+      KafkaSourceRDD.OffsetRange(tp, fromOffset, untilOffset, preferredLoc)
+    }.toArray
+
+    // Create a RDD that reads from Kafka and get the (key, value) pair as byte arrays.
+    val rdd = new KafkaSourceRDD(
+      sc, executorKafkaParams, offsetRanges, sourceOptions).map { cr =>
+        Row(cr.checksum, cr.key, cr.offset, cr.partition, cr.serializedKeySize,
+          cr.serializedValueSize, cr.timestamp, cr.timestampType.id, cr.topic, cr.value)
+    }
+
+    logInfo("GetBatch: " + offsetRanges.sortBy(_.topicPartition.toString).mkString(", "))
+    sqlContext.createDataFrame(rdd, schema)
+  }
+
+  /** Stop this source and free any resources it has allocated. */
+  override def stop(): Unit = synchronized {
+    consumer.close()
+  }
+
+  override def toString(): String = s"KafkaSource[$consumerStrategy]"
+
+  private def fetchPartitionOffsets(seekToLatest: Boolean): Map[TopicPartition, Long] = {
+    synchronized {
+      logTrace("\tPolling")
+      consumer.poll(0)
+      val partitions = consumer.assignment()
+      consumer.pause(partitions)
+      logDebug(s"\tPartitioned assigned to consumer: $partitions")
+      if (seekToLatest) {
+        consumer.seekToEnd(partitions)
+        logDebug("\tSeeked to the end")
+      }
+      logTrace("Getting positions")
+      val partitionToOffsets = partitions.asScala.map(p => p -> consumer.position(p))
```

Comment:
note to myself: this line will throw NPE when the topic is deleted just before it.
```
[info]   java.lang.NullPointerException:
[info]   at java.util.ArrayList.addAll(ArrayList.java:577)
[info]   at org.apache.kafka.clients.Metadata.getClusterForCurrentTopics(Metadata.java:257)
[info]   at org.apache.kafka.clients.Metadata.update(Metadata.java:177)
[info]   at org.apache.kafka.clients.NetworkClient$DefaultMetadataUpdater.handleResponse(NetworkClient.java:605)
[info]   at org.apache.kafka.clients.NetworkClient$DefaultMetadataUpdater.maybeHandleCompletedReceive(NetworkClient.java:582)
[info]   at org.apache.kafka.clients.NetworkClient.handleCompletedReceives(NetworkClient.java:450)
[info]   at org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:269)
[info]   at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.clientPoll(ConsumerNetworkClient.java:360)
[info]   at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:224)
[info]   at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:192)
[info]   at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:163)
[info]   at org.apache.kafka.clients.consumer.internals.Fetcher.listOffset(Fetcher.java:315)
[info]   at org.apache.kafka.clients.consumer.internals.Fetcher.resetOffset(Fetcher.java:298)
[info]   at org.apache.kafka.clients.consumer.internals.Fetcher.updateFetchPositions(Fetcher.java:170)
[info]   at org.apache.kafka.clients.consumer.KafkaConsumer.updateFetchPositions(KafkaConsumer.java:1409)
[info]   at org.apache.kafka.clients.consumer.KafkaConsumer.position(KafkaConsumer.java:1197)
```